src/wp-admin/includes/file.php

```diff
@@ -2557,3 +2627,65 @@ function wp_opcache_invalidate( $filepath, $force = false ) {
 
 	return false;
 }
+
+/**
+ * Attempts to clear the opcode cache for a directory of files.
+ *
+ * @since 6.2.0
+ *
+ * @see wp_opcache_invalidate()
+ * @link https://www.php.net/manual/en/function.opcache-invalidate.php
+ *
+ * @global WP_Filesystem_Base $wp_filesystem WordPress filesystem subclass.
+ *
+ * @param string $dir The path to the directory for which the opcode cache is to be cleared.
+ */
+function wp_opcache_invalidate_directory( $dir ) {
+	global $wp_filesystem;
+
+	if ( ! is_string( $dir ) || '' === trim( $dir ) ) {
+		if ( WP_DEBUG ) {
+			$error_message = sprintf(
+				/* translators: %s: The function name. */
+				__( '%s expects a non-empty string.' ),
+				'<code>wp_opcache_invalidate_directory()</code>'
+			);
+			// phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_trigger_error
+			trigger_error( $error_message );
+		}
+		return;
+	}
+
+	$dirlist = $wp_filesystem->dirlist( $dir, false, true );
+
+	if ( empty( $dirlist ) ) {
+		return;
+	}
+
+	/*
+	 * Recursively invalidate opcache of files in a directory.
+	 *
+	 * WP_Filesystem_*::dirlist() returns an array of file and directory information.
+	 *
+	 * This does not include a path to the file or directory.
+	 * To invalidate files within sub-directories, recursion is needed
+	 * to prepend an absolute path containing the sub-directory's name.
+	 *
+	 * @param array  $dirlist Array of file/directory information from WP_Filesystem_Base::dirlist(),
+	 *                        with sub-directories represented as nested arrays.
+	 * @param string $path    Absolute path to the directory.
+	 */
+	$invalidate_directory = function( $dirlist, $path ) use ( &$invalidate_directory ) {
+		$path = trailingslashit( $path );
+
+		foreach ( $dirlist as $name => $details ) {
+			if ( 'f' === $details['type'] ) {
+				wp_opcache_invalidate( $path . $name, true );
+			} elseif ( is_array( $details['files'] ) && ! empty( $details['files'] ) ) {
+				$invalidate_directory( $details['files'], $path . $name );
+			}
+		}
+	};
+
+	$invalidate_directory( $dirlist, $dir );
```

Comment:
Is it possible to avoid the closure via a recursive function (an actual question, not a passive aggressive change request)?

For this suggestion, I've changed the call to `::dirlist()` to be non-recursive that would be handled.

Sorry if I've missed a discussion already.

```suggestion
	$dir_list = $wp_filesystem->dirlist( $dir );

	if ( empty( $dir_list ) ) {
		return;
	}

	$path = trailingslashit( $dir );

	foreach ( $dir_list as $name => $details ) {
		if ( 'f' === $details['type'] ) {
			wp_opcache_invalidate( $path . $name, true );
		} elseif ( is_array( $details['files'] ) && ! empty( $details['files'] ) ) {
			wp_opcache_invalidate_directory( $path . $name );
		}
	}
```

Reply:
This would work but will call `$wp_filesystem->dirlist()` for each sub-directory. That's not needed as the first call returns the whole "tree". 

An alternative would be to pass the "current branch" from `$dir_list` as second argument to `wp_opcache_invalidate_directory()`, but then it makes that function's signature somewhat unexpected.

I know, closures are not great as they are virtually untestable, but they seem acceptable when the code it simple and easy to read/understand. Thinking this is one of these cases, just a foreach loop with a conditional :)

Reply:
@peterwilsoncc We had extensive discussion with @azaozz about this on the previous PR, https://github.com/WordPress/wordpress-develop/pull/3798

Reply:
👍 Thanks both. WFM.

Reply:
While this is resolved, I'll add this for our future selves and other contributors:

1. Closures can indeed make things testable.
2. However, this closure is essentially a very simple wrapper to recurse on `wp_opcache_invalidate()`, so testing its functionality would be, to some degree, testing `wp_opcache_invalidate()`.
3. Here's the branch coverage for the closure:
![image](https://user-images.githubusercontent.com/79332690/215933695-a93098bd-0fcc-4655-8371-15babc9aae30.png)